레포지토리 분리후 PR남깁니다.

iojkgh88 opened the pull request:
기존 리뷰 남겨주셨던부분은 따로 표시해두겠습니다.

iojkgh88 committed:
레포지토리 분리

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+@Controller
+public class UrlController {
+    private final UrlShorteningService urlShorteningService;
+
```
@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014826503)
@controller와 @RestController의 차이에 대해 알고 계신가용?

Author
@[gggorock](https://github.com/gggorock) gggorock [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014831537)
Controller 에 @responsebody 애노테이션이 추가된것입니다.
RestController로 지정해두지 않은 이유는 redirect보낼시에 viewResolver를 사용해야해서
redirect메소드를 제외하고 따로 ResponseBody를 붙였습니다.

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016124253)
아아 redirect:를 사용하려고 그러셨던거군요.
차이를 알고 계시다면 괜찮습니다. :)

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+    private Logger log = LoggerFactory.getLogger(getClass());
+
+    @PostMapping("/urls")
+    @ResponseBody
```
@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014827187)
@PostMapping("url") 경로 앞에는 /를 붙여주는게 관례입니다.
스프링 부트 3.0 버전으로 바뀌면 /를 붙이냐 붙이지 않냐에 따라 완전히 다른 경로로 인식하도록 바뀌게 되므로, 지금부터 잘 구분하는 습관을 가져보시면 좋을 것 같습니다

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014827706)
우리가 어떤 데이터를 URL에 붙여서 서버로 보내는지, 바디에 넣어서 보내는지는 무엇에 따라 결정하는게 좋을까요?
현재 세원님의 코드에서는 URL 뒤에 붙여서 보내고 있습니다. 일반적으로 이런 방식은 어떤 경우에 사용할까요?

Author
@[gggorock](https://github.com/gggorock) gggorock [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014841623) • 
경로관련 컨벤션이 있는지 몰랐습니다.
또한 3.0이후부터는 컨벤션 이상의 차이점이 생기는군요.
잘 구분해서 작성하겠습니다!

POST의 경우 자동으로 @RequestParam으로 하더라도 body로 받게되는것으로 배웠는데
포스트맨으로 확인해본 결과 body로 보내도 응답을 받고, url로도 응답이 보내지네요.

이전에 안전하지 않은 메서드의 경우 의도치않게 url접속을 하게될 수 있어 url로 접속하지 않도록 배웠는데 RequestBody로 명시적으로 받아야 할 것 같습니다.

변경하겠습니다!

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+    public AccessCountUrlDto generate(@RequestBody UrlDto urlDto) {
+        return  urlShorteningService.shorten(urlDto.getUrl());
+    }
+
```
@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014827456)
request.getServerName()+":"+request.getServerPort();

서버의 접속 정보를 백엔드 서버에서 다루는게 좋은 방식일까요?
그럼 혹시 프론트엔드 서버와 백엔드 서버가 서로 다른 서버에서 뜬다면 어떻게될까요?
그 때에도 동일한 코드로 동작할 수 있을지 생각해보시면 좋을 것 같습니다.

Author
@[gggorock](https://github.com/gggorock) gggorock [4 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1015451744) • 
아아 request.getLocalName() , request.getLocalPort으로 가져왔어야하네요!

프론트서버를 잘 몰라서 그런데, 프론트 페이지에서 클라이언트가 요청을 할때
클라이언트는 요청시 host를 프론트페이지로 보내나요?

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016126209) • 
서버랑 프론트엔드 페이지는 host 정보를 제외한 오직 path부터 그 뒷부분만 주고 받도록 만들면됩니다.
프론트엔드 페이지 내에서는 javascript를 통해 사용자가 지금 보고 있는 페이지의 host 정보를 뽑아낼 수 있습니다.
javascript로 개발자 도구에서 host 정보를 뽑아내는걸 한번 해보시죠. ㅎㅎ

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+    public AccessCountDto getUrl(@PathVariable("shortenUrl") String shortenUrl) {
+        return urlShorteningService.getUrlRequestCount(shortenUrl);
+    }
+
```
[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014827850) • 
여기서도 RequestParam을 사용하여 보내는 방식인데 이게 Restful한 API 설계일까요?

Author
@[gggorock](https://github.com/gggorock) gggorock [4 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1015103083) • 
"url/{index}" 형태로 pathVariable로 받아 처리해야합니다.

RequestParam으로 받는경우는 브라우저 의존적인 방식이며
formdata는 페이지를 넘기는 작업이 필요하기 때문에 비동기요청이 어렵습니다.

URI와 메소드만으로 의도된 동작이 가능하도록 처리해야합니다.

Form data 형식으로 테스트하던것과 더불어, RestApi에 대한 몰이해가 이러한 영향을 준것으로 보입니다.

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016138540)
넵! 한번 PathVariable 형태로 바꿔보시면 좋을 것 같습니다.
RequestParam을 사용하는 경우는 GET 메서드로 요청할 때 '검색의 조건'을 거는 경우 많이 사용되는 방식입니다.
특정 자원을 지칭하는 경우에는 PathVariable을 주로 사용하구요.
서비스들을 보면서 URL을 유심히 살펴보세요!

특히 쿠팡이 참고해보기 좋은 것 같습니다. ㅎㅎ

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+
+    //url 경로에 있는게 아니라서 컨트롤러를 분리해야하지 않을까 고민
+    @GetMapping("{shortenUrl}") // requestParam 사용 안하니 Controller에 지정된 요청 우선순위 해결. index.html보다 우선순위를 가져감 해결 필요
+    public String redirect(@PathVariable("shortenUrl") String shortenUrl) {
```
@[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014828069)
루트에 path변수를 지정하면 url, index.html보다 우선순위를 가져감 해결 필요

-> 이건 좋은 고민 포인트입니다. 해결할 수 있는 방법을 고민해보시면 좋을 것 같습니다. ㅎㅎ

Author
@[gggorock](https://github.com/gggorock) gggorock [4 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1015134761)
controller에 적용되지 않은 index.html만 다르게 처리된것으로 보입니다.
처음에 html form data 형식으로 index.html만 활용하다 Postman사용해보니 자원사용에는 이상없습니다.
index.html이 현재와서는 사용할 이유가 없어졌지만 방법을 찾아보겠습니다.

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -0,0 +1,49 @@
+    public String redirect(@PathVariable("shortenUrl") String shortenUrl) {
+        return "redirect:http://"+urlShorteningService.redirect(shortenUrl);
+    }
+
```
@[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014828524)
여기에 AOP를 사용하는건 조금 과할 수도 있을 것 같습니다. 아직 한번만 등장했으니까요.
이런게 반복된다면 말씀하신대로 AOP를 적용해볼 수도 있어요.

Author
@[gggorock](https://github.com/gggorock) gggorock [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014832742)
앗 넵.
log4j 처음 써보고나서 log위치에 따라 접근횟수 카운트가 반영이 안되다보니 return 뒤에 log가 적용되었으면 좋겠다 싶어서 적어두었습니다.

f-lab-fi reacted with thumbs up emoji
Author
@gggorock gggorock 4 days ago
추가질문입니다!
redirect처리시 service에서 url 객체의 접속카운트를 1올리기 때문에 안전하지 않은 메서드가 됩니다.
이경우에 GET으로 요청하는것이 적절하지 않아보이는데 POST로 보내는것이 맞을까요?

@f-lab-fi f-lab-fi 3 days ago
아닙니다.
그렇게하면 단축 URL 서비스로서 기능할 수가 없어집니다. 단순히 링크로 접속했을 때 이동할 수 없으니까요.
안전한 메서드와 멱등성 있는 메서드에 대해 자세히 찾아보시면, '자원의 상태가 바뀌는 것'을 기준으로 안전한지 멱등성 있는지를 정의합니다.
그럼 여기서 우리가 생각해봐야하는 것은 count도 과연 단축 URL의 상태라고 이야기할 수 있을까 입니다.
저는 count는 단축 URL의 상태라고 생각하기는 어렵다고 봅니다.
예를들어 우리가 게시판 서비스를 만들었을 때 게시글의 조회수를 게시글의 상태라고 이야기할 수 있을까요?
그런식이라면 게시글을 조회하는 API 역시 GET 방식이 아니라 POST가 되었어야 할 겁니다.
게시글에서 우리가 일반적으로 '자원의 상태'라고 이야기하는 것은 게시글의 제목이나 내용이지, 조회수까지 상태로 삼지 않습니다.

이 이야기가 주관적인 개념으로 들려서 혼동이 되실 수도 있지만, 혼동되시는게 맞습니다.
자원의 상태는 정의하기 나름이에요. 다만, 우리가 비즈니스 로직과 비즈니스 로직이 아닌 로직을 나누는 것처럼, 자원에서 어떤 값이 핵심적인지 생각해보시면 좋을 것 같습니다. :)

Author
@[gggorock](https://github.com/gggorock) gggorock [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016122573)
자원의 상태에서 상태의 범위는 모든 상태라기 보다는,
클라이언트에게 변경될시 의사결정이 변경될정도로 중요한 요소로 생각해보아도 될것같네요.

이 서버에서 GET요청을 변한다고 했을때, 자원의 상태범위는 단축URL의 변경여부이고,
조회수는 의도적으로 GET요청을 많이한다해도 실제 중요한 요소인 단축URL에는 문제가 발생하지 않으니 안전하다고 판단하면 될것같습니다!

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlShorteningService.java:
```diff
@@ -0,0 +1,82 @@
+        return urlRepository
+                .findById(id)
+                .orElseThrow(RuntimeException::new);
+    }
```
[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014828823)
왜 id를 저장하나 했는데, 단축 URL을 id를 인코딩하여 만들고 계시는군요.
그런데 이러면 id가 몇인가에 따라 단축 URL이 조금씩 길어지지 않을까요~?

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014829342)
이건 크게 코드를 수정해야할텐데, id가 없어도 되는 서비스로 바꿔보시면 좋을 것 같습니다.

Author
@[gggorock](https://github.com/gggorock) gggorock [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014836456)
id로 되어있지만 index와 값이 같고 Repository의 index를 기준으로 반환합니다.
ArrayList가 아니라 다른 자료형을 썼을경우도 고려해서 index를 id로 명명했는데
어차피 ArrayList로 만드는건데

UrlShortenService에서 shorten메서드에서 URL인스턴스를 새로 만드는데
그것을 indexOf(Url)로 가져오면 O(n)으로 성능이 떨어질것 같아서 메모리를 포기하고 index를 Url의 Id로 인스턴스에 저장하고 getId로 가져오는 방식을 택했습니다.

ArrayList의 index만 사용하는 형태로도 구현해보겠습니다!

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016120305)
넵 그렇게 바꿔보시면 좋을 것 같습니다. :)

@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [3 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1016121039)
우선 너무 성능을 고려하진 않으셔도 좋을 것 같아요.
이 이후에 바꿔볼 과제는 Map을 사용하는 방식과 List를 사용하는 방식이 같은 인터페이스를 통해 서비스 되도록 만드는겁니다.

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlShorteningService.java:
```diff
@@ -0,0 +1,82 @@
+                    Url newUrl = new Url(normalUrl);
+                    urlRepository.save(newUrl);  // save반환을 index로한 이점을 살리고 싶었는데 익명클래스의 블록은 블록바깥변수를 참조하지 못하네요.(인텔리제이 자동추천은 AtomicInteger) 그래서 index를 Url도메인클래스의 변수로 두고 꺼내쓰기로 했습니다.
+                    return newUrl;
+                });
```
[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014829068)
save반환을 index로한 이점을 살리고 싶었는데 익명클래스의 블록은 블록바깥변수를 참조하지 못하네요.(인텔리제이 자동추천은 AtomicInteger) 그래서 index를 Url도메인클래스의 변수로 두고 꺼내쓰기로 했습니다.

네 방향성은 적절한지 생각해봐야겠지만, 지금 상태에서는 잘 하신겁니다.

iojkgh88 commented on src/main/java/com/gggorock/urlshortener/domain/UrlShorteningService.java:
```diff
@@ -0,0 +1,82 @@
+                });
+
+        String shortenUrl = "/"+ConverterWithBase62.encode(urlRepository.indexOf(url));
+
```
[f-lab-fi](https://github.com/f-lab-fi) [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014829228)
여기 등장하는 hostName는 위에서 코멘트 드렸던 것처럼 서버에서 처리되는게 적절한지 생각해보시기 바랍니다. :)

iojkgh88 commented on src/main/resources/static/index.html:
```diff
@@ -0,0 +1,17 @@
+<form action="/url" method="post">
+     <input type="text" name="url">
+    <input type="submit" value="압축하기!">
+</form>
```
@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014829656)
여유 되시면, ajax를 통해 페이지 갱신 없이 비동기적으로 단축 URL 생성을 요청하고, 단축된 URL이 페이지에 나오도록 바꿔보시면 좋을 것 같습니다. :)
분명 UI/UX에 대한 시각을 키우시는 계기가 될거에요.
UI/UX에 대한 고려 없이 좋은 API가 나올 수 없습니다.

Author
@[gggorock](https://github.com/gggorock) gggorock [4 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1015136268)
아, 이거는 꼭 해보고싶은데 다른것들 하기 바빠서 조금 미뤄두고 있습니다.
반드시 시도해보겠습니다!

iojkgh88 commented on src/test/java/com/gggorock/urlshortener/domain/UrlEncoderWithBase62Test.java:
```diff
@@ -0,0 +1,32 @@
+        int test = 15324235;
+        String encode = urlEncoderWithBase62.encode(test);
+        System.out.println("encode = " + encode);
+
```
@[f-lab-fi](https://github.com/f-lab-fi) f-lab-fi [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014829874)
테스트는 원래 성공, 실패 여부가 나와야 하는데 이런 테스트는 예외가 던져지는게 아니라면 테스트 실패 여부를 확인할 수 가 없습니다. 아래쪽에 decode에서 디코딩 되는 테스트를 하긴 하지만, encode 그 자체로도 동작의 성공 여부가 테스트가 되어야합니다.

Author
@[gggorock](https://github.com/gggorock) gggorock [5 days ago](https://github.com/gggorock/flab-mentoring-records/pull/2#discussion_r1014837550)
encode가 Index를 인코딩해서 값을 변환해주는것인데 인코딩된것을 어떻게 Base62로 인코딩한것이 맞는지 적절한 테스트방법을 찾지 못해서 변환되는지만 간단하게 봤었습니다.

생각해보니 직접계산해본 값을 넣어보는것으로 진행했어도 되었을것 같습니다.

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/ConverterWithBase62.java:
```diff
@@ -0,0 +1,55 @@
+            }
+        }
+        return -1;
+    }
```
BASE62_TABLE에 없으면, -1이 아니라 예외가 던져져야 하지 않을까요~?

iojkgh88 replied:
아, 만들때 인코딩된 데이터만 들어갈거라고 생각하고 알고리즘문제풀이처럼 이상있는경우 인덱스를 -1로 반환하게했네요.

정적메서드에 public이다보니 얼마든지 다른 값이 들어갈 수 있겠고 그 값이 -1이라면, decode값에도 영향을 미치고, 외부에서도 그 의미를 곡해해서 쓸수있을것 같다는 생각이 듭니다.

말씀하신대로 예외를 던지는 것이 맞는 것 같습니다!

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/ArrayListUrlRepository.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public synchronized int save(Url url) { // ArrayList의 스레드안정성이 문제가되는 경우로 동기화(urlList의 인덱스와 Url의 인덱스의 정합성이 모두 고려되어 메소드 전체 동기화)
+            int index = -1;
+            urlList.add(url);
```
int형 index를 기준으로 전체 애플리케이션이 돌아가고 있는데요, 혹시 int형 index 없이 애플리케이션이 돌아가도록 만들어보실 수 있을까요?

iojkgh88 replied:
네 index없이 작성해보았습니다!

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/Url.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final String originUrl;
+
+    private Long requestCount = 0L;
```
https://github.com/gggorock/url-shortener/pull/1/files#r1020221761

에도 쓴 것 처럼, index를 중심으로 애플리케이션이 돌아가고 있습니다. 이건 아마도 Url에 단축된 URL이 저장되지 않고 있는 것 같아서 처럼 보입니다. 단축 URL에서 핵심적인 데이터는 **단축된 URL -> 원래의 URL**을 매핑하는겁니다. 그런데 지금 코드에선 원래라면 Url이 가지고 있어야할 단축된 URL 정보가 Repository 쪽으로 세어나가고 있어보여요.

단축된 URL을 Url 클래스 안으로 가져오면서, index를 사용하지 않고 코드를 작성해보셔야 이런 문제를 해결할 수 있을겁니다.

iojkgh88 replied:
도메인 객체가 어떤 상태값을 가져야하는지 생각하게 된 부분입니다. 
한번의 해결이 있었으면 도메인이 상태를 품고 도메인을 재사용하도록 하는것이 효율적이라는 생각이 들었습니다. 
단순히 필드를 아낀다고 득이되지 않는군요

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/ConverterWithBase62.java:
```diff
@@ -0,0 +1,55 @@
+        StringBuilder sb = new StringBuilder();
+        int quotient = index;
+        do {
+            sb.append(BASE62_TABLE[(int) (quotient % 62)]);
```
`sb.append(BASE62_TABLE[(int) (quotient % 62)]);`
이 코드 잘 읽히시나요? 가독성에 문제가 있어보입니다. 너무 괄호가 많아요.

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/dto/AccessCountDto.java:
```diff
@@ -0,0 +1,28 @@
+public class AccessCountDto {
+    private final int index;
+    private final String originUrl;
+    private final Long accessCount;
```
`private final int index;`를 클라이언트에게 전달하는게 의미가 있나요..?
이러면 base62 뿐만 아니라 클라이언트가 index까지 알게되는건데, 불필요해보입니다.

iojkgh88 replied:
이것이야말로 캡슐화를 전혀 신경 안쓴 코드네요.
혹시나 알려주면 클라이언트에게 뭔가 도움이 되겠지 수준의 생각으로 넘겼습니다. ㅠ

iojkgh88 committed:
2nd revision
1. 인덱스에 강한 의존관계구조 변경
2. Dto 변환위치를 Service로 명확화
3. UrlIn dto객체의 유효성검사 추가
4. Base62UrlConverter 코드 가독성 증대
5. ArrayList 저장소 외 Map 저장소도 구현
6. 주요 어플리케이션 계층 클래스를 package-private로 변경

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/Base62UrlConverter.java:
```diff
@@ -1,27 +1,33 @@
-            sb.append(BASE62_TABLE[(int) (quotient % 62)]);
+            int digitIndex = (int) (quotient % 62); // 아 이부분을 좀더 가독성있게 표현하기가 어렵네요.
+            char digit = BASE62_TABLE[digitIndex];
+            sb.append(digit);
```
지금 정도 나누셨으면 괜찮습니다~

iojkgh88 replied:
넵!

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/UrlController.java:
```diff
@@ -19,31 +23,32 @@ public class UrlController {
+    public ShortenUrlOut generate(@RequestBody @Valid UrlIn urlDto, BindingResult bindingResult) throws BindException {
+        if (bindingResult.hasErrors()) {
+            throw new HttpClientErrorException(HttpStatus.BAD_REQUEST);
+        }
```
컨트롤러에 대한 유효성을 검사하는 부분을 추가하셨네요.
잘 바꿔보셨습니다.
여기서 하나 더 찾아보시면 좋을만한 내용이 있는데요, 우리가 유효성 검사를 해줄 때 이렇게 컨트롤러에서 유효성 검사를 해주기도 하지만, 도메인 객체에서도 유효성 검사를 하기도 합니다. 이렇게 2가지 경우가 어떤 차이가 있는건지 찾아보시면 좋을 것 같아요.

iojkgh88 replied:
오류페이지에서는 400에러라고 하는데 
json으로 나가는 오류코드는 500으로 나가서 사실 반쪽짜리 유효성검사이긴 합니다.
그리고 라이브러리에서 제공하는 @URL 은 scheme만 확인하다보니 {"url":"https://"} 라고만 보내도 통과가 되더군요.
유효성검사 커스텀방법 및 에러코드를 따로 해소해야 합니다.

dto에 대한 유효성검사와 도메인객체에서 유효성검사의 차이를 알아보겠습니다.

iojkgh88 replied:
컨트롤러의 유효성 검사는 클라이언트 측에서 들어오는 값의 형식과, 타입을 검증합니다. 
도메인 객체는 도메인이 갖는 값이 비즈니스로직에 부합하는지 여부를 검증합니다. 

제가 작성하고 있는 이 서비스에 빗대어 보면,
제가 컨트롤러에서 진행하고자 한 유효성 검사는 들어오는 UrlIn dto라는 형식을 갖춘가운데(이것은 Spring에서 자동적으로 처리) url이 url주소인지에 대한 타입검증목적입니다.
도메인객체(Url)의 경우는 예외를 던지진 않고 있지만, 이 서비스에서는 중복된 단축 URL이 존재할 수 없고, 중복된 원본주소값은 같은 단축 URL이어야 한다는 점을 보장할 때만 도메인 객체가 생성되고 해당 값을 변경하지 못합니다. 

https://blog.frankdejonge.nl/where-does-validation-live/

f-lab-fi replied:
넵 잘 찾아보셨네요~!
나중에 프로젝트 할 때에도 컨트롤러에서 진행하는 유효성 검사와 도메인 객체에서 진행하는 유효성 검사를 나눠서 진행해보시기 바랍니다.

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/MapUrlRepository.java:
```diff
@@ -0,0 +1,40 @@
+@Repository
+class MapUrlRepository implements UrlRepository {
+
+    private final Map<String, Url> store = new ConcurrentHashMap<>();
```
Map을 사용한 레포지토리를 추가해보니 List 쪽 레포지토리랑 인터페이스를 맞출 수 없는 부분이 생기진 않으셨나요?
일단 index를 없애는 과정에 묻혀 그 차이가 느껴지지 않았을지도 모르겠지만, 순수하게 List쪽 레포지토리를 구현한 후 Map쪽 레포지토리를 추가하는 과정에서 인터페이스(public method)가 변경되어야 했다면, 최초에 인터페이스가 잘못 설계된 겁니다. List에 의존적인 인터페이스였던거죠.

iojkgh88 replied:
말씀하신 포인트대로 index를 없애는 과정에서 map에 유연하게 대응될수있는 interface로 변경했기 때문에 빠른 변경이 가능했다고 적어두었습니다. 
다만 interface의 변경이 잦은부분에서 문제의식은 크게 갖고있었습니다.

실질적으로는 연역적으로 설계가 되었다기 보다는 귀납적으로 맞춰지게 되어버렸습니다. 

수정했던 이유는 아래와 같습니다.

1. repository에서 CRUD를 Service계층에서 사용되는 메서드들이 어떤것인지 명확히 그려지지 않아, service를 구현해가면서  repository의 인터페이스에서 필요한부분과 필요하지 않은 부분을 계속 수정
 - save()를 만들 때 반환타입을 Url와  int로 고민하다 Service 구현과정에서 유효한 것으로 수정했습니다.
 - update()메서드가 사용되지 않아 삭제
 - 모든 단축URL 접속수를 내보내는 서비스를 위해 findAll()이 추가
 - Index걷어낸 이후로는 Index대신 shortenUrl을 사용하여 findByShortenUrl() 추가 

2. Index를 걷어내는 과정에서의 변경
- 리뷰받기 전에는 findById() 를 잘못이해해서 findByIndex로 변경 
- 도메인객체 필드에 shortenUrl를 추가하게 되어 findByIndex() 삭제

interface를 변경하면서 느낀것이 코드작성 경험이 중요하구나 느꼈습니다.

파이님께서는 최근 interface설계시 어떤점을 생각하시면서 작업하셨는지 궁금합니다.

f-lab-fi replied:
소프트웨어 개발은 경험이 중요합니다. 지금 나와있는 프로그래밍 방법론, 디자인 패턴 등은 전부 귀납적으로 도출된 결과합니다.
세원님이 해보신 과정 자체가 지극히 자연스러운 경험입니다.

인터페이스 설계를 어떻게 하는지에 대해서 물어보셨는데, 저 역시 제가 작성하는 인터페이스에 대한 귀납적으로 도출된 결과부터 활용합니다.
예를들면 [Repository 인터페이스를 두는 것 역시 하나의 디자인 패턴](https://deviq.com/design-patterns/repository-pattern)처럼 자리잡고 있습니다.
이렇게 패턴화된 인터페이스는 해당 패턴을 최대한 따라가도록 작성합니다.
물론 그 외에 패턴화되지 않은 인터페이스들을 정의해야할 일도 생깁니다. 주로 도메인 객체들에 대한 인터페이스(인터페이스라는 표현은 public method 들을 포함합니다.)를 정의할 때는 최대한 표현력 있는 메서드를 만들려고 노력하고, 객체 내부적으로 하는 일에 대한 캡슐화가 잘 이루어지도록 설계하려고 합니다. :)

f-lab-fi commented on src/main/java/com/gggorock/urlshortener/domain/UrlFactory.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final UrlConverter urlConverter;
+    private final AtomicLong sequence = new AtomicLong(0L);
+    private final long VOID_URL_SEQUENCE;
```
ㅎ.. index가 어디로 숨었나 했더니 여기로 와서 sequence가 됐군요.
일단 바꿔주신 부분은 잘 바꾸셨습니다. repository 쪽에 복잡하게 들어와 있던 부분이 해결됐으니까요.
다만, sequence 자체도 없애는 것을 의도하긴 했습니다. sequence를 사용하지 않고도 단축 URL을 생성할 수 있는 방법은 없을까요? 그냥 랜덤한 문자열을 반환한다던지?

iojkgh88 replied:
sequence값을 이용하지 말라고 하셨던 의도는 일종에 단축URL생성시 규칙이 밖으로 노출되는 문제 때문일까요? 

계속 고민해보고 있는데 양의 long범위의 랜덤값이라 하더라도 중복여부를 어떻게하면 효율적으로 처리할지 생각중입니다.
배열에 생성된 값을 저장해서 하는것은 메모리 낭비이고, 기존 레포지토리에 있으면 다시 랜덤값을 반환하는 식으로 작성해보고자 합니다. (ArrayList는 조금 느릴수도 있을것같은데 Map이라면 성능개선이 가능할것같습니다!)

iojkgh88 replied:
위와 같은 방법으로 재 구현 성공했습니다!(일단 리뷰에서 요청하셨던 부분 전체가 처리되면 올리도록 하겠습니다.)
랜덤값으로 받으니 그럴듯한 단축열주소가 나오네요. 

구현시 class로 구현되었던 UrlFactory를 인터페이스로 바꾸고
기존 Sequence를 사용하던 Factory를 SequenceUrlFactory로 바꾸고,
Random 값을 이용한 Factory를 구현하여 의존성주입을 진행했습니다.

얼추 Factory Method Pattern비슷하게 나오네요!
기존 Service나 Repository와 같은 코드를 수정하지 않고 확장에 성공했습니다!

iojkgh88 replied:
<img width="482" alt="image" src="https://user-images.githubusercontent.com/91076824/201522184-6182c24d-d133-4b25-8a24-84958ffeb05c.png">

f-lab-fi replied:
> sequence값을 이용하지 말라고 하셨던 의도는 일종에 단축URL생성시 규칙이 밖으로 노출되는 문제 때문일까요?

-> 그런 측면도 고려해볼 수 있긴 하지만, 실제 애플리케이션이 서비스 되는 상황을 고려해서 이야기드린 부분입니다. 만약 서버가 여러개가 떠서 서비스한다면 지금 코드로도 동작할 수 있을까요? 아마 그때는 다시 Repository 쪽으로 코드를 옮겨서 데이터베이스 같은 수단을 통해 sequence를 관리하려고 하겠죠. 그럼 이 sequence가 전체 시스템의 병목이 될 수 있습니다. 그런 측면에서 영향을 낮춰보는 방법에 대해 고민해보시라고 추천드렸습니다. :)

iojkgh88 replied:
아아 stateless한 설계를 의도하셨던 것이군요!
피상적으로 스케일아웃을 위해 무상태를 유지하는것이 중요하다고 들었는데, 
sequence가 상태를 유지하도록 만드는 원흉이었네요.

f-lab-fi replied:
> 얼추 Factory Method Pattern비슷하게 나오네요!
> 기존 Service나 Repository와 같은 코드를 수정하지 않고 확장에 성공했습니다!

-> 잘 해보셨습니다. 다른 코드를 수정하지 않고 확장할 수 있던건 잘 추상화 되어있기 떄문입니다. :)
앞으로도 새로운 요구사항이 생겼을 때 클라이언트쪽 코드를 수정하지 않고 확장할 수 있도록 객체지향적인 코드를 작성해보세요.

f-lab-fi commented on src/test/java/com/gggorock/urlshortener/domain/UrlShorteningServiceTest.java:
```diff
@@ -1,24 +1,23 @@
-import static org.junit.jupiter.api.Assertions.*;
-
 class UrlShorteningServiceTest {
 
```
테스트 관련해서 몇가지 더 해보시면 좋을 만한 것을 추천드리겠습니다.

1. 예외가 던져지는 상황에 대한 테스트
2. 현재 작성한 것처럼 통합 테스트 외에 단위 테스트 작성. 예를들어 Repository를 제외하고 UrlShorteningService만을 테스트 할 수 있는 단위 테스트
3. 컨트롤러에 대해 HTTP 요청을 하는 것처럼 테스트 (MockMvc 사용)

iojkgh88 replied:
테스트 어떻게 해야할지 막막했는데 가능한 전략을 찾을수 있도록 제시해주셔서 감사합니다!

1. jupiter api Assertions에 assertThrows() 메서드 사용
2. Mock 객체는 구현체가 없어도 기능을 명세해주는것만으로 단위테스트를 가능하게 해주네요! 신기하네요. 
3. MockMvc는 Mock 객체의 역할을 대략 이해하니 Http신호를 가짜로 구현해주는 라이브러리로 생각됩니다. 

관련 라이브러리들 공부하면서 업데이트 해보겠습니다.
